pilot/pkg/model/push_context.go

```diff
@@ -1707,11 +1715,27 @@ func (ps *PushContext) mergeGateways(proxy *Proxy) *MergedGateway {
 				workloadLabels = labels.Collection{proxy.Metadata.Labels}
 			}
 			if workloadLabels.IsSupersetOf(gatewaySelector) {
-				out = append(out, cfg)
+				selected = true
 			}
 		}
-	}
+		if selected {
+			for _, s := range gw.Servers {
+				if servicePort, ok := targetPorts[s.Port.Number]; ok && servicePort != s.Port.Number {
+					// Check if the gateway server port is also defined as a service port, if so skip rewriting since it is
+					// ambiguous on whether the server port points to service port or target port.
+					if _, ok := servicePorts[s.Port.Number]; ok {
+						continue
+					}
 
+					// The gateway server is defined with target port. Convert it to service port before gateway merging.
+					// Gateway listeners are based on target port, this prevents duplicated listeners be generated when build
+					// listener resources based on merged gateways.
+					s.Port.Number = servicePort
```

Comment:
I think you are mutating it for all references. need a deep copy probably

Reply:
we do similar with VS when we resolve hosts to FQDN

Reply:
yeah, it is mutating the k8s informer cache, which is not encouraged

Reply:
Done.